lib/models/updateAction.js

```diff
@@ -86,6 +130,67 @@ function doIt(action, event, callback) {
         });
 }
 
+function makeTokenListenerFunc(action, event, callback) {
+    return function tokenHandlerFunc(error, token) {
+        logger.debug('tokenHandlerFunc %s %s', error, token);
+        if (error || !token) {
+            return callback(error);
+        } else {
+            logger.debug('tokenHandlerFunc retrying with', token);
+            return doRequest(action, event, token, function cbDoReqUpdAxn(error, data) {
+                callback(error, data);
+            });
+        }
+    };
+}
+
+function doItWithToken(action, event, callback) {
+    var service,
+        tenant,
+        ruleName,
+        cached,
+        newTokenListener;
+
+    service = event.service;
+    tenant = event.tenant;
+    ruleName = event.ruleName;
+    newTokenListener = makeTokenListenerFunc(action, event, callback);
+
+    cached = getCachedToken(tenant, service, ruleName);
+    if (cached.token === null) {  // First time token is generated from trust
+        cached.emitter.once(newTokenEventName, newTokenListener);
+        if (cached.generating === false) {
+            logger.debug('generating token for %s %s %s', tenant, service, ruleName);
+            generateToken(action.parameters.trust, cached);
+        }
+    } else if (cached.generating === true) { // In the middle of getting a new one
+        logger.debug('waiting token to be generated %s %s %s', tenant, service, ruleName);
+        cached.emitter.once(newTokenEventName, newTokenListener);
+    } else {
+        doRequest(action, event, cached.token, function cbDoReqUpdAxn(error, data) {
+            logger.debug('cbDoReqUpdAxn %s %s', error, data);
+            // If authorization error, we should call generateToken and addListener to retry
```

Comment:
The description of the behaviour could be the following:

"First time an update action using trust token is triggered, Perseo interacts with Keystone to get the temporal auth token corresponding to that trust token. This auth token is cached and used in every new update associated to the same action. Eventually, Perseo can receive a 401 Not Authorized due to auth token expiration. In that case, Perseo interacts again with Keystone to get a fresh auth token, then retries the update that causes the 401 (and the cache is refreshed with the new auth token for next updates). 

It could happen (in theory) that a just got auth token also produce a 401 Not authorized, however this would be an abnormal situation: Perseo logs the problem with the update but doesn't try to get a new one from Keystone. Next time Perseo triggers the action, the process may repeat, i.e. first update attemp fails with 401, Perseo requests a fresh auth token to Keystone, the second update attemp fails with 401, Perso logs the problem and doesn't retry again."

(By the way, I'm not sure if 401 is the error code got in this case; please, adjust the text with the right one in that case).

In addition, regarding documentation, I understand that Architecture document should be updated to included Keystone stuff. At least the "External components that interact with Perseo", "External data flows" and "Scenarios" sections (the later maybe just adding a new diagram "Update action with trust token" or something like that).